### PR TITLE
i#2001 trace perf: elide rip-rel and same-base addrs

### PR DESCRIPTION
**Pull request by derekbruening**
Reduces drcachesim offline tracing overhead by eliding rip-relative
and same-unmodified-base addresses from the recorded trace,
reconstructing their values during post-processing.  In measurements
this is pretty significant, removing 12%-17% of entries that need to
be written out during tracing.

Adds identification of elidable memory operands to instru_offline_t,
exported both to the runtime tracer and the raw2trace post-processor.

Changes raw2trace's instruction cache key to a pair <tag,pc> to handle
tail-duplicated blocks.  Adds elision identification through a
constructed instrlist when first encountering a block tag.  Adds a new
struct memref_summary_t to store elision information with each cached
memory operand.

Increases the offline file version and adds versioning support for
backward compatibility with no-elision trace, as well as to make it
easier to keep compatibility when more elision cases are added in the
future.

Adds a file type to the offline file header to identify filtered
traces as a sanity check and to avoid extra work when there are no
elided addresses at all.

Adds a statistics interface to retrieve raw2trace metrics.  The
initial metric is the number of elided addresses.

Issue: #2001

### CODE REVIEW

File: clients/drcachesim/tests/raw2trace_io.cpp [hunk 1]
```diff
@@ -201,7 +201,8 @@ test_trace_timestamp_reader(const raw2trace_directory_t *dir)
     file->read((char *)buffer, BUFFER_SIZE_BYTES(buffer));
 
     std::string error;
-    if (!trace_metadata_reader_t::is_thread_start(buffer, &error) && !error.empty())
+    if (!trace_metadata_reader_t::is_thread_start(buffer, &error, nullptr, nullptr) &&
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tests/raw2trace_io.cpp]: Not sure what you mean by "whitespace reformat": two arguments were added and there is a change here to pass those in?

[the remark this answers is not in the document]
Comment (Carrotman42) [reply clients/drcachesim/tests/raw2trace_io.cpp]: Yep, just me misreading the code in front of me. Pardon the noise.

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 2]
```diff
@@ -619,33 +621,75 @@ raw2trace_t::do_conversion()
...
-                               app_pc orig)
+raw2trace_t::get_instr_summary(void *tls, uint64 modidx, uint64 modoffs,
+                               app_pc block_start, INOUT app_pc *pc, app_pc orig)
 {
     auto tdata = reinterpret_cast<raw2trace_thread_data_t *>(tls);
     const app_pc decode_pc = *pc;
     // For rep string loops we expect the same PC many times in a row.
-    if (decode_pc == tdata->last_decode_pc)
+    if (decode_pc == tdata->last_decode_pc) {
+        *pc = tdata->last_summary->next_pc();
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tracer/raw2trace.cpp]: Ah yes this is a general bugfix where the instr processing loop would get the same instr multiple times in a row.  With this PR that turned into something fatal.  Without this fix it should still cause a problem (just not visible).  Should I separate it out?

**Comment by derekbruening** [comment]
The CI failures are real: compiler warnings on Windows (VS2013 doesn't support some C++11 stuff like "constexpr" and "noexcept" -- hopefully we can complete the full switch to VS2017 and drop VS2013 soon); actual bug revealed 32-bit which unfortunately was overdoing the elision so the actually safe to elide savings may be lower than previously measured (this is push-pop addr sharing: was planning a separate PR to do that so we'd get the savings eventually, just not in this PR); and a non-x86 issue.  Working on them.

The push-pop bug makes me want to have a better test.  Unfortunately we want to test both the tracer and post-proc, which means a real app (not just a checked-in or synthetic trace), which adds non-determinism.  Maybe an allasm app with as many corner cases as we can think of?  Or should we add an option to turn off all the opts (older disp elision, addr elision here) and turn off ASLR and have a single-threaded app that we run twice and hope there's no other source of differences between the runs??

**Comment by derekbruening** [comment]
(The test I'm talking about adding is comparing the precise addresses in the final trace which I didn't make clear in the prior post.)

**Comment by derekbruening** [comment]
The AArch64 failure was from exclusive store instrumentation being *after* the instr instead of before.  The elision checker only looked for a label at the insertion point instead of the target app instr.

**Comment by derekbruening** [comment]
burst_malloc is also failing but that's a separate issue: #3974.

File: clients/drcachesim/tracer/instru_offline.cpp [hunk 3]
```diff
@@ -550,4 +592,124 @@ offline_instru_t::bb_analysis(void *drcontext, void *tag, void **bb_field,
...
+void
+offline_instru_t::identify_elidable_addresses(void *drcontext, instrlist_t *ilist,
+                                              int version)
+{
+    // Analysis for eliding redundant addresses we can reconstruct during
+    // post-processing.
+    // We can't elide when doing filtering.
+    if (memref_needs_full_info)
+        return;
+    instr_t *instr;
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tracer/instru_offline.cpp]: Done.

File: clients/drcachesim/tracer/raw2trace.h [hunk 4]
```diff
@@ -103,6 +147,47 @@ struct instr_summary_t final {
...
+
+    /**
+     * Sets properties of the "pos"-th destination memory operand by OR-ing in the
+     * two boolean values.
+     */
+    void
+    set_mem_dest_flags(size_t pos, bool use_remembered_base, bool remember_base)
+    {
+        DEBUG_ASSERT(num_mem_srcs_ + pos < mem_srcs_and_dests_.size());
+        mem_srcs_and_dests_[num_mem_srcs_ + pos].use_remembered_base =
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tracer/raw2trace.h]: +1.  Done.  (Since this is C++, technically `|=` should be safe: it just doesn't feel right though and I'd be worried somebody would then model some of our C code on it...)

File: clients/drcachesim/tracer/raw2trace.h [hunk 5]
```diff
@@ -1023,6 +1318,8 @@ class raw2trace_t : public trace_converter_t<raw2trace_t> {
         std::istream *thread_file;
         std::ostream *out_file;
         std::string error;
+        int version;
+        offline_file_type_t file_type;
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tracer/raw2trace.h]: +1.  This was my mistake to not initialize it.

File: clients/drcachesim/tracer/raw2trace.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Carrotman42) [reply clients/drcachesim/tracer/raw2trace.cpp]: No need to separate it out IMO, can you just add a line somewhere in the description along the line of this?

"Also fix a bug in raw2trace_t::get_instr_summary where the INOUT parameter 'pc' was not always correctly updated."

[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tracer/raw2trace.cpp]: Yes, done (put into new commit which will then show up in the edit box).

**Comment by derekbruening** [comment]
Well, it took a lot of effort to add this new test, and to get it working on all platforms.  It required fixing some little issues on other platforms.  It might be nice to separate some of those out but there are yet more issues and other tracker entries cover the rest.  With this test we should have more confidence in this elision PR.

PTAL

File: clients/drcachesim/tests/burst_traceopts.cpp [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+                                      STDERR, true, true);
+                if (memref1.data.pc != memref2.data.pc) {
+                    disassemble_with_info(dcontext,
+                                          reinterpret_cast<byte *>(memref2.data.pc),
+                                          STDERR, true, true);
+                }
+                error_string = "Data fields do not match";
+                return false;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tests/burst_traceopts.cpp]: I will add a comment along the lines of "We check fields which trace optimizations affect: the core instruction and data fetch entries.  The current optimizations have no impact on other entries, and many other entries have variable values such as timestamps from run to run."

File: clients/drcachesim/tests/burst_traceopts.cpp [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+
+/* Asm routines. */
+extern "C" {
+void
+test_disp_elision();
+void
+test_base_elision();
+};
+
+class trace_comparer_t : public analysis_tool_t {
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tests/burst_traceopts.cpp]: Hmm.  On the one hand it is logically a trace analysis tool, which is why I wrote it like one, but OTOH this exposed iterator lets us just directly call anything we want, so you're right that it can just be a function.  It does have some state, with a counter for better diagnostics on mismatches, but that can be solved.  Sure, will do.

**Comment by derekbruening** [comment]
> I was a little surprised to see the new code in arm to support standalone, but I guess if you want the test to support all platforms then it's necessary. I'm not sure the best way to get this committed, I'm happy to think through it with you.

I have split off 4 PR's: #3985, #3986, #3987, #3988

**Comment by derekbruening** [comment]
I am also hitting this on Appveyor:
```
	c:\projects\dynamorio\suite\tests\api\ir_x86.c(463) : fatal error C1002: compiler is out of heap space in pass 2
```

**Comment by derekbruening** [comment]
PTAL.  I've merged in the 4 separate PR's so those parts of the patch are gone.

**Comment by derekbruening** [comment]
The Appveyor failure seems to be specific to Windows 8.1/Server 2012 WOW64.  I had to build a custom diagnostics circular buffer logging into raw2trace to figure out what is going on since I can't reproduce it locally:

```
ERROR: failed to initialize analyzer: raw2trace failed: Failed to process
file for thread 68: memref entry found outside of bb

----------------------------------------------
Appending 3 instrs in bb 016C0680 in mod 12 +0x40680 = C:\windows\SYSTEM32\ntdll.dll
  0x778d0680 cld
  0x778d0681 mov    0x04(%esp)[4byte] -> %ecx
Appended memref type 0 size 4 to 006FF2D8
  0x778d0685 mov    (%esp)[4byte] -> %ebx
Filling in elided addr with remembered esp: 006FF2D4
Appended memref type 0 size 4 to 006FF2D4
Appending 4 instrs in bb 016C0685 in mod 12 +0x40685 = C:\windows\SYSTEM32\ntdll.dll
  0x778d0685 mov    (%esp)[4byte] -> %ebx
Filling in elided addr with remembered esp: 00000000   <=========== PROBLEM!
Appended memref type 0 size 4 to 00000000
  0x778d0688 push   %ecx %esp -> %esp 0xfffffffc(%esp)[4byte]
Appended memref type 1 size 4 to 006FF2D0
  0x778d0689 push   %ebx %esp -> %esp 0xfffffffc(%esp)[4byte]
Appended memref type 1 size 4 to 006FF2D0
  0x778d068a call   $0x778fd16e %esp -> %esp 0xfffffffc(%esp)[4byte]
Appended memref type 1 size 4 to 006FF2CC
----------------------------------------------

Raw input:
----------------------------------------------
0x2006001800040680
0x00000000006ff2d4
0x2008001800040685
0x00000000006ff2d4
0x00000000006ff2d4
0x00000000006ff2d0
0x00000000006ff2cc
----------------------------------------------
```

Looks like it's re-using the stored base from the end of the prior bb.
Why does the prior end there anyway, and then the next starts with a
duplicate?  That doesn't seem right.  And I deliberately put into raw2trace
lookup by the pair of block-tag,pc to avoid a tail-duplication issue, even
if this was the correct block sequence.  So how is the 2nd one using the
same elision data?

I don't have that ntdll.  But a similar one from Win8.1 shows the problem,
by searching for "cld" followed by that particular memref:

```
_KiUserExceptionDispatcher@8:
  6B2C0CC0: FC                 cld
  6B2C0CC1: 8B 4C 24 04        mov         ecx,dword ptr [esp+4]
  6B2C0CC5: 8B 1C 24           mov         ebx,dword ptr [esp]
  6B2C0CC8: 51                 push        ecx
  6B2C0CC9: 53                 push        ebx
  6B2C0CCA: E8 DF D0 02 00     call        _RtlDispatchException@8
```

So it's a Ki hook.  Our trampoline and displaced code is being interpreted in a strange way.  Are we playing games with the start pc?  It still doesn't seem like you'd end up w/ the symptoms above.  I will try to get more data on the block sequence.

**Comment by derekbruening** [comment]
Travis failures are: #3496 drmgr-test and #3974 burst-malloc
